File: euler45.cpp
#include <iostream>
#define max 100000
using namespace std;


bool inarr(long *arr, int len, long val) 
{
	for (int i=0;i<len;i++) {
		if(arr[i]==val)return true;
		if(arr[i]>val) return false;
	}
	return false;
}

int main()
{
	long a;
	a=(long)400000*(3*400000-1)/2;
	cout << a<<endl<<endl;
	long penta[max], tri[max], hexa[max];
	for (int i=1; i<=max;i++){
		penta[i-1]=(long)i*(3*i-1)/2;
		hexa[i-1]=(long)i*(2*i-1);
		tri[i-1]=(long)i*(i+1)/2;
	}
	int cnt=0;
	for(int i=0;i<max;i++){
		long c=tri[i];
		if(inarr(penta, max, c)){
			if(inarr(hexa, max, c)) {cnt++; if(cnt==3){cout << c<< endl;return 0;}}
		}
	}
}

File: euler45.py
# just solve the quadratic equation

def solveQuadratic(a,b,c):
    if b**2 < 4*a*c:return 0
    return (float(-b) + (b**2-4*a*c)**0.5)/float(2*a)

def genTriangle(n):
    return n*(n+1)/2

def genPentagonal(n):
    return n*(3*n-1)/2

def genHexagonal(n):
    return n*(2*n-1)

n = 1
cnt = 0

while cnt<3:
    triangle = n*(n+1)/2
    penta = solveQuadratic(3, -1, -2*triangle)
    hexa = solveQuadratic(2, -1, -triangle)
    if genPentagonal(int(penta))==triangle and genHexagonal(int(hexa))==triangle:
        cnt+=1
        print triangle
    n+=1
